api/src/main/java/org/openmrs/api/impl/CohortServiceImpl.java

```diff
@@ -57,108 +59,160 @@ public Cohort saveCohort(Cohort cohort) throws APIException {
 		return dao.getCohortByUuid(uuid);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#addPatientToCohort(org.openmrs.Cohort,
 	 *      org.openmrs.Patient)
 	 */
 	public Cohort addPatientToCohort(Cohort cohort, Patient patient) {
-		if (!cohort.contains(patient)) {
-			cohort.getMemberIds().add(patient.getPatientId());
+		if (!cohort.contains(patient.getPatientId())) {
+			CohortMembership cohortMembership = new CohortMembership(patient);
+            Context.getCohortService().addMembershipToCohort(cohort, cohortMembership);
 			Context.getCohortService().saveCohort(cohort);
 		}
 		return cohort;
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#removePatientFromCohort(org.openmrs.Cohort,
 	 *      org.openmrs.Patient)
 	 */
 	public Cohort removePatientFromCohort(Cohort cohort, Patient patient) {
-		if (cohort.contains(patient)) {
-			cohort.getMemberIds().remove(patient.getPatientId());
+		if (cohort.contains(patient.getPatientId())) {
+            cohort.removeMember(patient, "Removing patient from Cohort");
 			Context.getCohortService().saveCohort(cohort);
 		}
 		return cohort;
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getCohortsContainingPatient(org.openmrs.Patient)
 	 */
 	@Transactional(readOnly = true)
 	public List<Cohort> getCohortsContainingPatient(Patient patient) {
 		return dao.getCohortsContainingPatientId(patient.getPatientId());
 	}
-	
+
 	@Transactional(readOnly = true)
 	public List<Cohort> getCohortsContainingPatientId(Integer patientId) {
 		return dao.getCohortsContainingPatientId(patientId);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getCohorts(java.lang.String)
 	 */
 	@Transactional(readOnly = true)
 	public List<Cohort> getCohorts(String nameFragment) throws APIException {
 		return dao.getCohorts(nameFragment);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getAllCohorts()
 	 */
 	@Transactional(readOnly = true)
 	public List<Cohort> getAllCohorts() throws APIException {
 		return Context.getCohortService().getAllCohorts(false);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getAllCohorts(boolean)
 	 */
 	@Transactional(readOnly = true)
 	public List<Cohort> getAllCohorts(boolean includeVoided) throws APIException {
 		return dao.getAllCohorts(includeVoided);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getCohort(java.lang.String)
 	 */
 	@Transactional(readOnly = true)
 	public Cohort getCohort(String name) throws APIException {
 		return dao.getCohort(name);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#purgeCohort(org.openmrs.Cohort)
 	 */
 	public Cohort purgeCohort(Cohort cohort) throws APIException {
 		return dao.deleteCohort(cohort);
 	}
+
+	/**
+	 * @see org.openmrs.api.CohortService#addMembershipToCohort(org.openmrs.Cohort,
+	 * 		org.openmrs.CohortMembership)
+	 */
+	public Cohort addMembershipToCohort(Cohort cohort, CohortMembership cohortMembership) throws APIException {
+        cohort.addMembership(cohortMembership);
```

Comment:
You need to throw an exception here if the cohort is voided, i believe this might be a clue as to why that test that has been bothering you is failing, I think you are adding a patient to a voided cohort so when you call getCohortsContainingPatient() it never gets returned this cohort since the query excludes voided cohorts